recalibrate() fix

yarocoder opened the pull request.

bplimley commented:
Please update your branch from `bearing:master` since there have been several changes in `bearing:master` that are not in your fork yet.

bplimley commented:
@yarocoder you mentioned at the meeting you had updates, could you please push so that they appear here? Thanks.

laurac8r commented:
Hi Brian,

Yeah, it prevented me from making an edit to the code last night without fixing the discrepancies between my commit and the current working `calibrate.py` code.

Let me know if you want any edits to fix the mismatching metadata error with `Gamma_Analysis.py` (possible solutions indicated in my comment for the associated issue in the Issues tab on this repository).

yarocoder commented:
Hi Brian,

The recent commits are for both the `recalibrate()` and Sea Salt issues. The changes I've made are as follows:

**`calibrate.py`:** In the `recalibrate()` function, the `_recal` identifier is removed from every element containing it from essentially the `sample_measurements` object (called `files` in `recalibrate()`) before being passed into the `calibration_table` function. Also, I've fixed the capitalization in the check for `_recal.Spe` which may or may not have been an issue in the original code. Now I think the functionality is solid for the `recalibrate()` function, but a new issue has appeared: A "peak not found" error is issued by `calibration_check()`. I need help with this error because the peak finding is a more technical aspect of the code, and it's part of the changes that need to be made (see **Required Changes** below).

**`Gamma_Reference.py`:** Renamed the `Reference()` class into `base_reference_class()`. Created an inherited class `petri_reference_class()` that adds the specific count rate as an attribute `ref_spec_ct_rate`, while preserving the reference mass, concentration, concentration error, and conversion attributes. Created a specific count rate list `petri_specific_count_rate` that is pulled from elements N2-N19 from the `UCB018 Loc010 (reference) Petri` Google sheet. Created a list of samples names that require the use of the Petri reference. Created a separate soil mass for the Petri reference and renamed the old S5F soil mass reference. Changed the use of the S6F soil mass to the S5F soil mass as taken from the associated Google Sheets. Created a `petri_reference` attribute of `Gamma_Reference.py` that creates an instance of `petri_reference_class()` and inputs the Petri reference mass, concentration, concentration error, conversion, and specific count rate array.

**`Gamma_Analysis.py`:** In the `main()` function, assigned the S5F reference by default. Added a check if the sample uses the Petri reference, if so it assigns the specific activity attribute of `petri_reference` in place of the `UCB018_Soil_Sample010_2.Spe` spectrum. In the `analyze_isotope()` method, added a check if the reference is of type `base_reference_class()` or `petri_reference_class()` and if so assigns the `soil_reference` or `petri_reference` attributes of `Gamma_Reference.py` as the reference object for mass, concentration, concentration error, and conversion (and specific count rate in the case of the Petri reference). Added the same check when the code finds the reference peak, peak area, emission, emission uncertainty, and activity. The usual method is used if the reference object is the S5F reference class, and a simple assignment of the `ref_spec_ct_rate` attribute of `petri_reference` is done if the reference object is the Petri reference class. Either way, the `reference_activity` and `sample_comparison` objects are input into the `isotope_concentration` method. In `isotope_concentration`, I've added the check of reference objects, but there is no functionality (see **Required Changes** below). One additional change is the removal of the input of the detector efficiency into `emission_rate()` function and the use of it in that function. We should keep the `absolute_efficiency()` function in case we need it in the future, but currently the code does not need it.

**Required changes:** In `Gamma_Reference.py`, there needs to be a list of the which isotopes each element of `petri_specific_count_rate` refers to and an attribute for an isotope list can be added to it. This can play into the idea of having a dict for the isotope references in `isotope_concentration()`. I think the major issue in `analyze_isotope()` is that it calls on `isotope_activity()` to find a weighted average of the isotope activity based on the emission rate array input into it, which becomes a problem for the Petri reference since we only have the specific count rate for each peak. So I need help breaking the `petri_specific_count_rate` list based on which isotope each peak energy belongs to and help with either creating an emission rate array for the Petri reference or finding the weighted average of the specific count rate for each isotope in the Petri reference and inputting that into the `isotope_concentration()` function.

laurac8r commented:
See comment in Sea Salt pull request!

bplimley commented on calibrate.py:
```diff
@@ -164,24 +173,29 @@ def recalibrate(files):
+    #         files.remove(file)
+
+    files_updated = [os.path.splitext(files[i])
+                     [0].replace("_recal", "")
```
Let's leave "recal" in the filenames for the calibration table. That way it is clear which file was analyzed.

laurac8r replied:
The problem is that all "recal" files are removed from `cal_offsets` so `calibration_table()` crashes since the filenames and offsets arrays are inherently not the same length, which is what `calibration_table` checks for.

bplimley commented on calibrate.py:
```diff
@@ -164,24 +173,29 @@ def recalibrate(files):
+
+    files_updated = [os.path.splitext(files[i])
+                     [0].replace("_recal", "")
+                     for i in np.arange(len(files))]
```
Also, no need to index here - you can just do `[os.path.splitext(f)[0] for f in files]` or so

laurac8r replied:
Ok, yeah thanks for the help Brian, that cleans things up quite a bit!

bplimley commented on calibrate.py:
```diff
@@ -177,9 +177,8 @@ def recalibrate(files):
-    files_updated = [os.path.splitext(files[i])
-                     [0].replace("_recal", "")
-                     for i in np.arange(len(files))]
+    files_updated = [os.path.splitext(f)[0].replace("_recal", "")
```
This variable is only used in the `calibration_table` call below. The `replace` is not changing the length of the file list, it's only changing the name of the files printed in the table. I'm saying the name in the table should be the `_recal` version, not the base version. The `_recal` version is what is analyzed, right?

bplimley commented:
The changes remaining in this pull request are either already included in #60, or they are reversions. So I'm closing this now.
